triggers/daml/AutoApproval.daml

```diff
@@ -0,0 +1,132 @@
+  custodyRoles              <- queryByField @CustodyRole.Role        @"provider" (.provider) (== party)
+  let custodyRequests  = generateRoleCommands (\cid -> CustodyRole.AcceptCustodyRequest with custodyRequestCid = cid) custodyServicesRequests custodyRoles
+      issuanceRequests = generateRoleCommands (\cid -> CustodyRole.AcceptIssuanceRequest with issuanceServiceRequestCid = cid) issuanceServiceRequests custodyRoles
+
+  auctionServiceRequests    <- queryByField @AuctionService.Request @"provider" (.provider) (== party)
+  biddingServiceRequests    <- queryByField @BiddingService.Request @"provider" (.provider) (== party)
+  distributionRole          <- queryByField @DistributionRole.Role  @"provider" (.provider) (== party)
+  let auctionRequests = generateRoleCommands (\cid -> DistributionRole.ApproveAuctionServiceRequest with auctionServiceRequestCid = cid) auctionServiceRequests distributionRole
+      biddingRequests = generateRoleCommands (\cid -> DistributionRole.ApproveBiddingServiceRequest with biddingServiceRequestCid = cid) biddingServiceRequests distributionRole
+
+  tradingServiceRequests    <- queryByField @TradingService.Request @"provider" (.provider) (== party)
+  listingServiceRequests    <- queryByField @ListingService.Request @"provider" (.provider) (== party)
+  tradingRole               <- queryByField @TradingRole.Role       @"provider" (.provider) (== party)
+  let tradingRequests = generateRoleCommands (\cid -> TradingRole.ApproveTradingServiceRequest with tradingRequestCid = cid) tradingServiceRequests tradingRole
+      listingRequests = generateRoleCommands (\cid -> TradingRole.ApproveListingServiceRequest with listingRequestCid = cid) listingServiceRequests tradingRole
+
+  submitCommands $
+    custodyRoleOffers ++
+    custodyRoleRequests ++
+    custodyServicesOffers ++
+    custodyRequests ++
+    distributionRoleOffers ++
+    distributionRoleRequests ++
+    auctionRequests ++
+    biddingRequests ++
+    issuanceServiceOffers ++
+    issuanceRequests ++
+    listingServiceOffers ++
+    regulatorServiceOffers ++
+    settlementServiceOffers ++
+    settlementServiceRequests ++
+    tradingRoleOffers ++
+    tradingRoleRequests ++
+    matchingserviceOffers ++
+    matchingserviceRequests
+
+-- | Facilitates querying a template by a specific field which matches a provided filter predicate
+queryByField : forall t f a m. (Template t, HasField f t a, ActionTriggerAny m, Functor m) => (t -> a) -> (a -> Bool) -> m [(ContractId t)]
+queryByField getField pred = map fst . filter (pred . getField . snd) <$> query
+
+-- | Generates commands in the expected format expected by the trigger for a contract / choice combination
+generateCommands : forall t c r. (Template t, Choice t c r) => c -> [(ContractId t)] -> [(AnyContractId, Command)]
+generateCommands choiceArg queryResults = foldl (\acc cid -> (toAnyContractId cid, exerciseCmd cid choiceArg) :: acc) [] queryResults
```

Comment:
I'm curious, why use `foldl` here instead of `map`, unless you prefer the resulting list to be in reverse?

Reply:
`foldl` makes the code look more complex and thus it gives the general impression that I know what I'm actually doing 😄 

Joking aside, the reason is that initially this function was much more complicated that warranted a `foldl` but since I've removed that logic, I could have also replaced the `foldl` with a `map` to make the entire function simpler. I'll update it in the next PR.

Reply:
Haha fair enough! Related, I can't tell you how many times I've written what I thought was a beautiful and elegant recursive function that I realize later could have just been a simple fold...

Reply:
Argh, I could have levelled up in complexity by writing my own custom recursive function. I must remember that for next time...

In the meantime, I've refactored out `foldl` with `map` on the last commit 👍